File: src/leetcode/0125-Valid-Palindrome/two-pointers.ts
function isPalindrome(s: string): boolean {
  const REGEX = /[^A-Za-z0-9]/g // ^ inside [...] negates the class
  const cleanedStr = s.toLowerCase().replace(REGEX, "")

  let start = 0
  let end = cleanedStr.length - 1

  while (start <= end) {
    if (cleanedStr[start] !== cleanedStr[end]) return false
    start++
    end--
  }

  return true
}

console.log(isPalindrome("A man, a plan, a canal: Panama")) // "amanaplanacanalpanama" → true
console.log(isPalindrome("race a car")) // "raceacar" → false
console.log(isPalindrome(" ")) // "" → true

File: src/lcpy/leetcode/0125-valid-palindrome/two_pointers.py
import re


class Solution:
    REGEX = re.compile(r"[^a-zA-Z0-9]")

    def isPalindrome(self, s: str) -> bool:
        str = self.REGEX.sub("", s.lower())
        right, left = 0, len(str) - 1
        while right < left:
            if str[right] != str[left]:
                return False

            right += 1
            left -= 1

        return True


sol = Solution()
print(sol.isPalindrome("A man, a plan, a canal: Panama"))  # true
print(sol.isPalindrome("race a car"))  # false
print(sol.isPalindrome("race e car"))  # true
print(sol.isPalindrome(" "))  # true
